internal/controller/postgrescluster/autogrow.go

```diff
@@ -0,0 +1,189 @@
+			// If there was an error parsing the value, treat as unset (equivalent to zero).
+			desiredRequestBackup = ""
+			previous, _ = resource.ParseQuantity("")
+
+		}
+	}
+
+	// determine if the appropriate volume limit is set
+	limitSet := limitIsSet(cluster, volumeType, instanceSetName)
+
+	if limitSet && current.Value() > previous.Value() {
+		r.Recorder.Eventf(cluster, corev1.EventTypeNormal, "VolumeAutoGrow",
+			"%s volume expansion to %v requested for %s/%s.",
+			volumeType, current.String(), cluster.Name, instanceSetName)
+	}
+
+	// If the desired size was not observed, update with previously stored value.
+	// This can happen in scenarios where the annotation on the Pod is missing
+	// such as when the cluster is shutdown or a Pod is in the middle of a restart.
+	if desiredRequest == "" {
+		desiredRequest = desiredRequestBackup
+	}
+
+	return desiredRequest
+}
+
+// limitIsSet determines if the limit is set for a given volume type and returns
+// a corresponding boolean value
+func limitIsSet(cluster *v1beta1.PostgresCluster, volumeType, instanceSetName string) bool {
+
+	var limitSet bool
+
+	switch volumeType {
+
+	// Cycle through the instance sets to ensure the correct limit is identified.
+	case "pgData":
+		for _, specInstance := range cluster.Spec.InstanceSets {
+			if specInstance.Name == instanceSetName {
+				limitSet = !specInstance.DataVolumeClaimSpec.Resources.Limits.Storage().IsZero()
+			}
+		}
+	}
+	// TODO: Add cases for pgWAL and repo volumes
+
+	return limitSet
+
+}
+
+// setVolumeSize compares the potential sizes from the instance spec, status
+// and limit and sets the appropriate current value.
+func (r *Reconciler) setVolumeSize(ctx context.Context, cluster *v1beta1.PostgresCluster,
+	pvc *corev1.PersistentVolumeClaim, volumeType, instanceSpecName string) {
+
+	log := logging.FromContext(ctx)
+
+	// Store the limit for this instance set. This value will not change below.
+	volumeLimitFromSpec := pvc.Spec.Resources.Limits.Storage()
+
+	// Capture the largest volume size currently defined for a given instance set.
+	// This value will capture our desired update.
+	volumeRequestSize := pvc.Spec.Resources.Requests.Storage()
+
+	// A limit of 0 is ignorned, so the volume request is used.
+	if volumeLimitFromSpec.IsZero() {
+		return
+	}
+
+	// If the request value is greater than the set limit, use the limit and issue
+	// a warning event.
+	if volumeRequestSize.Value() > volumeLimitFromSpec.Value() {
+		r.Recorder.Eventf(cluster, corev1.EventTypeWarning, "VolumeRequestOverLimit",
+			"%s volume request (%v) for %s/%s is greater than set limit (%v). Limit value will be used.",
+			volumeType, volumeRequestSize, cluster.Name, instanceSpecName, volumeLimitFromSpec)
+
+		pvc.Spec.Resources.Requests = corev1.ResourceList{
+			corev1.ResourceStorage: *resource.NewQuantity(volumeLimitFromSpec.Value(), resource.BinarySI),
+		}
+		// Otherwise, if the limit is not set or the feature gate is not enabled, do not autogrow.
```

Comment:
you moved the "limit not set logic" right? I think this comment needs to be updated?